lib/request.js

```diff
@@ -0,0 +1,11 @@
+export const formatRequestData = (request, response, env) => {
+  const url = new URL(request.url);
+  const timestamp = Date.now();
+  const originResponse = response || {};
+  const apiKey = request.headers.get('api-key') || url.searchParams.get('api-key') || (request.headers.get('Authorization')?.startsWith('Bearer ') ? request.headers.get('Authorization').substring(7) : 'unknown');
+  const cfCache = (originResponse) ? (response.headers.get('CF-Cache-Status') || 'miss') : 'miss'
+  const formattedUrl = url.toString().replaceAll('=', '\\=');
+
+  // We're setting field value to 1 to count the number of requests
+  return `${env.INFLUX_METRIC},status_code=${originResponse.status},url=${formattedUrl},hostname=${url.hostname},pathname="${url.pathname}",method=${request.method},cf_cache=${cfCache},api_key=${apiKey} value=1 ${timestamp}`
```

Comment:
```suggestion
  // https://docs.influxdata.com/influxdb/cloud/reference/syntax/line-protocol/
  return `${env.INFLUX_METRIC},status_code=${originResponse.status},url=${formattedUrl},hostname=${url.hostname},pathname="${url.pathname}",method=${request.method},cf_cache=${cfCache},api_key=${apiKey} value=1 ${timestamp}`
```

Reply:
We also need to talk about the data scheme:

Here we're setting a bunch of things as tags (status_code, url, api_key, etc). In InfluxDB, indexes are created over tags. That means that if you have unbounded values in tags, the index space quickly explodes.

Therefore, anything that is unbounded (like `url` and `api_key` for example) needs to be a field instead of a tag. When we move things from tags to fields we can also remove the `value` field.

Reply:
Didn't know that, thanks for pointing that up. I'm going to move those tag fields into value fields.

Do you think that current request fields we have are enough or we'd need to add/remove anything?

Reply:
Is `pathname` enough to identify an endpoint? It works now because all spark-stats endpoints use the GET method. We will likely forget about this when adding the first non-GET endpoint to spark-stats.

What do you think about adding `method` to the list of fields? IMO, it's easier to do it now while we are deep at working with this codebase.

Reply:
I'm proposing to for now drop pathname _and_ method, as we're only interested in the per-key usage. We can look up in Cloudflare which endpoint is used how much. I believe we don't yet need api-key x pathname stats

Reply:
I am fine with both options - either drop or collect - both pathname and method together.

I feel we have reached the point where we are spending more time discussing the change than it would cost to implement it. Let's decide and move on.

Reply:
@juliangruber If we move `api_key` to value we won't be able to group by as it's only possible to group by tags or time in influxdb https://docs.influxdata.com/influxdb/cloud/query-data/influxql/explore-data/group-by/

If we're worried about value being unbound we can either limit it's length or hash the values before sending them to influxdb.

Reply:
We are using Flux to query InfluxDB. I believe we can group using values when using Flux, see https://docs.influxdata.com/influxdb/cloud/query-data/flux/group-data/

> With Flux, you can group data by any column in your queried data set. “Grouping” partitions data into tables in which each row shares a common value for specified columns. This guide walks through grouping data in Flux and provides examples of how data is shaped in the process.
>
> (...)
> 
> Flux’s [group() function](https://docs.influxdata.com/flux/v0/stdlib/universe/group) defines the group key for output tables, i.e. grouping records based on values for specific columns.
>
> ```flux
> dataStream
>   |> group(columns: ["cpu", "host"])
> ```

If that does not work, then here is another direction we can explore:
1. Write raw data to InfluxDB, storing the `api_key` as a value.
2. Write another observer component (similar to [observeRetrievalResultCodes](https://github.com/filecoin-station/spark-stats/blob/15a67a94c59007f3d9fda727b0cab4360adc2015/observer/lib/observer.js#L100C14-L100C41) to periodically fetch the raw data from InfluxDB, aggregate it (with one-day granularity?) and write the aggregated data to spark-stats' Postgres database, where the REST API server can read it from.

Reply:
I have tried that but It's not working properly. Maybe I'm doing something wrong here?

<img width="1356" alt="Screenshot 2024-12-12 at 12 50 46" src="https://github.com/user-attachments/assets/1e4ea144-e74b-479c-989e-a353ec796f0e" />

Reply:
AFAICT, the measurement is called `spark_stats_api_usage_local`.

<img width="861" alt="Screenshot 2024-12-12 at 14 06 47" src="https://github.com/user-attachments/assets/3e133223-4979-4f14-9273-0ad8260cbc68" />

Reply:
This is how far I got with the exploration. We need to find out how to aggregate & count the results for each api-key and the given time period (1 minute in my case).

```flux
from(bucket: "api-observability") 
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r._measurement == "spark_stats_api_usage_dev" and r._field == "api_key")
  |> group(columns: ["_value"])
  |> aggregateWindow(every: 1m, fn: count, createEmpty: false)
```

Reply:
This query seems to work for me.

```
from(bucket: "api-observability") 
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r._measurement == "spark_stats_api_usage_dev" and r._field == "api_key")
  |> duplicate(column: "_value", as: "api_key")
  |> group(columns: ["api_key"])
  |> aggregateWindow(every: 1h, fn: count)
```

<img width="864" alt="Screenshot 2024-12-12 at 15 22 32" src="https://github.com/user-attachments/assets/25b29169-7755-41ff-92d0-7b753857fe92" />

Reply:
@bajtos Awesome! Good job on figuring it out!